DataDog/datadog-agent
# [system-probe] Automatically download kernel headers (1)

### What does this PR do?

Leverages the `nikos` library to allow the system-probe to automatically download kernel headers in the event that 
a) runtime compilation is enabled, and 
b) kernel headers are not found on the host machine.  

To make reviewing a little easier, this PR contains only the changes needed for downloading kernel headers on non-RPM distributions. I will open a follow-up PR which will contain the changes necessary for building `nikos` with support for RPM distributions (using `omnibus-nikos`).

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

1. In your `system-probe.yaml` file, enable runtime compilation & set `kernel_header_dirs` to `/dev/null` (this tells the system-probe to check for kernel headers in an incorrect location & forces the download of kernel headers):
```
system_probe_config:
  enabled: true
  log_level: debug
  enable_runtime_compiler: true
  kernel_header_dirs: /dev/null
```
2. Start the system-probe & verify in the logs that kernel headers get downloaded/validated, and that runtime compilation of the network tracer succeeds:
```
...
| Downloading kernel headers for target distribution Ubuntu, release 20.04, kernel 5.4.0-58-generic
...
| successfully downloaded kernel headers to [/var/tmp/datadog-agent/system-probe/kernel-headers/lib/modules/5.4.0-58-generic/build]
| found valid kernel headers at /var/tmp/datadog-agent/system-probe/kernel-headers/lib/modules/5.4.0-58-generic/build
...
| module: network_tracer started
...
```
3. In a separate window, check that the tracer telemetry indicates the correct outcome:
```
$ curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq .compilationTelemetryByAsset.tracer

{
  "runtimeCompilationEnabled": true,
  "runtimeCompilationResult": "CompilationSuccess",
  "runtimeCompilationDuration": <some number>,
  "kernelHeaderFetchResult": "DownloadSuccess"
}
```
4. Delete the runtime-compiled asset(s):
```
sudo rm -rf /var/tmp/datadog-agent/system-probe/build
```
5. Restart the system-probe & check that once again runtime compilation succeeded, this time by using the previously downloaded headers:
```
$ curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq .compilationTelemetryByAsset.tracer

{
  "runtimeCompilationEnabled": true,
  "runtimeCompilationResult": "CompilationSuccess",
  "runtimeCompilationDuration": <some number>,
  "kernelHeaderFetchResult": "DownloadedHeadersFound"
}
```
